# Урок 1. Расчет вероятности случайных событий

In [1]:
from math import factorial

In [2]:
# перестановка P - последовательное расположение с учетом порядка
def P(n):
    return factorial(n)

# размещение A - перестановка k элем-ов из множ-ва n
def A(n, k):
    return factorial(n) // factorial(n - k)

# сочетание C - кол-во способов взять k элем-ов из множ-ва n без учета порядка расположения
def C(n, k):
    return factorial(n) // (factorial(k) * (factorial(n - k)))

***
### Задание 1. Из колоды в 52 карты извлекаются случайным образом 4 карты.
### a) Найти вероятность того, что все карты – крести.

Воспользуемся формулой подсчета количества сочетаний по $k$ элементов из множества $n$: \
$C_n^k=\cfrac{n!}{k!(n-k)!}$

И классической формулой вероятности: \
$P(A)=\cfrac{m}{n}$ , где $m$ - число **благоприятных** исходов, $n$ - общее число **всех** исходов.

В колоде из 52 листов $\cfrac{52}{4}=13$ карт одной масти.

Тогда в нашем случае \
$n = C_{52}^{4}$ - количество способов извлечь 4 карты из 52 \
$m = C_{13}^{4}$ - способов извлечь 4 карты одной масти

In [3]:
m = C(13,4)
print(f'{m = }')

n = C(52,4)
print(f'{n = }')

p = m / n
print(f'P(4трефы) = {m} / {n} = {p:.4f}')

m = 715
n = 270725
P(4трефы) = 715 / 270725 = 0.0026


Из классической формулы вероятности следует, что вероятность извлечь \
первую крестовую: $\cfrac{13}{52}$; вторую крестовую: $\cfrac{12}{51}$; третью крестовую: $\cfrac{11}{50}$; четвертую крестовую: $\cfrac{10}{49}$

Тогда $P(4 трефы) = \cfrac{13}{52} \cdot \cfrac{12}{51} \cdot \cfrac{11}{50} \cdot \cfrac{10}{49}$

In [4]:
p = 13/52 * 12/51 * 11/50 * 10/49
print(f'P(4трефы) = {p:.4f}')

P(4трефы) = 0.0026


### б) Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

В этом случае так же общее число исходов $n = C_{52}^{4}$, а $m$ будет складываться из 4 вариантов:
1. из 4-х карт 1 туз: \
$C^1_4$ - способов достать 1 туз из 4-х и \
$С^3_{48}$ - способов достать три остальные карты, тогда \
$C^1_{4} \cdot С^3_{48}$ - общее количество сочетаний 1 туз и 3 не туза.<br><br>
2. из 4-х карт 2 туза: \
$C^2_4$ - способов достать 2 туза из 4-х и \
$С^2_{48}$ - способов достать две остальные карты, тогда \
$C^2_{4}\cdotС^2_{48}$ - общее количество сочетаний 2 туза и 2 не туза.<br><br>
3. из 4-х карт 3 туза: \
$C^3_4$ - способов достать 3 туза из 4-х и \
$С^1_{48}$ - способов достать оставшуюся карту, тогда \
$C^3_{4}\cdotС^1_{48}$ - общее количество сочетаний 3 туза и 1 не туз.<br><br>
4. из 4-х карт 4 туза: \
$C^4_4$ - способов достать 4 туза из 4-х.

Общее число благоприятных исходов - сумма этих 4х вариантов.

In [5]:
m = C(4,1)*C(48,3) + C(4,2)*C(48,2) + C(4,3)*C(48,1) + C(4,4)
print(f'{m = }')

n = C(52,4)
print(f'{n = }')

p = m / n
print(f'P(хотя бы 1 туз) = {m} / {n} = {p:.4f}')

m = 76145
n = 270725
P(хотя бы 1 туз) = 76145 / 270725 = 0.2813


Так же можно использовать то, что сумма вероятностей противоположных событий равна $1$, т.е. $P(A)+P(\bar{A})=1$.\
$\bar{A}$ - событие, когда из 4-х извлеченных карт не оказалось ни одного туза. \
Для этого события общее число исходов $n = C_{52}^{4}$, а $m = C_{48}^{4}$ - количество способов извлечь 4 карты из колоды **без** тузов. \
Тогда $P(A)=1-P(\bar{A})$

In [6]:
m = C(48,4)
print(f'{m = }')

n = C(52,4)
print(f'{n = }')

p = 1 - m / n
print(f'P(хотя бы 1 туз) = 1 - {m} / {n} = {p:.4f}')

m = 194580
n = 270725
P(хотя бы 1 туз) = 1 - 194580 / 270725 = 0.2813


#### Ответ:
а) $P(\text{4 трефы})\approx 0.0026 \approx 0.26\%$
<br>
б) $P(\text{хотя бы 1 туз})\approx 0.2813 \approx 28.13\%$

***
### Задание 2. На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9. Код содержит три цифры, которые нужно нажать одновременно. Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

Так как кнопки нажимаются **одновременно**, то порядок не важен: \
$n=C_{10}^{3}$ - общее число исходов = количиство сочетаний 3-х элементов из 10, \
$m = 1$ - количество благоприятных исходов (замок открылся).

In [7]:
m = 1
print(f'{m = }')

n = C(10,3)
print(f'{n = }')

p = m / n
print(f'P(открыть с первой попытки) = {m} / {n} = {p:.4f}')

m = 1
n = 120
P(открыть с первой попытки) = 1 / 120 = 0.0083


#### Ответ:
$P(\text{открыть с первой попытки})\approx 0.0083 \approx 0.83\%$

### Задание 3. В ящике имеется 15 деталей, из которых 9 окрашены. Рабочий случайным образом извлекает 3 детали.  Какова вероятность того, что все извлеченные детали окрашены?

$n = C_{15}^{3}$ - общее количество исходов = количество сочетаний 3 деталей из 15, \
$m = C_{9}^{3}$ - благоприятных исходов = количеством сочетаний 3-х окрашенных деталей из 9.

In [8]:
m = C(9,3)
print(f'{m = }')

n = C(15,3)
print(f'{n = }')

p = m / n
print(f'P(3 из 3х окрашены) = {m} / {n} = {p:.4f}')

m = 84
n = 455
P(3 из 3х окрашены) = 84 / 455 = 0.1846


Или $P(\text{3 из 3х окрашены})=\cfrac{9}{15} \cdot \cfrac{8}{14} \cdot \cfrac{7}{13}$

In [9]:
p = 9/15 * 8/14 * 7/13
print(f'P(3 из 3х окрашены) = {p:.4f}')

P(3 из 3х окрашены) = 0.1846


#### Ответ:
$P(\text{3 из 3х окрашены})\approx 0.1846 \approx 18.46\%$

***
### Задание 4. В лотерее 100 билетов. Из них 2 выигрышных. Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

$\displaystyle P(\text{оба выигрышных})=\frac{1}{C^2_{100}}$<br>

In [10]:
m = 1
print(f'{m = }')

n = C(100,2)
print(f'{n = }')

p = m / n
print(f'P(оба выигрышных) = {m} / {n} = {p:.4f}')

m = 1
n = 4950
P(оба выигрышных) = 1 / 4950 = 0.0002


Или $\displaystyle P(\text{оба выигрышных})=\frac{2}{100}\cdot\frac{1}{99}$

In [11]:
p = 2/100 * 1/99
print(f'P(оба выигрышных) = {p:.4f}')

P(оба выигрышных) = 0.0002


#### Ответ:
$P(\text{оба выигрышных})\approx 0.0002 \approx 0.02\%$
___